### PR TITLE
Support for Dockerized infrastructure server

### LINKED ISSUE #415: Ad-hoc packaging of leastauthority.com infrastructure server creates many maintenance challenges
Here are some examples of issues that seem to stem from the current approach: #293 #314 #281 #262 #259 #254

Packaging the infrastructure server using a more sophisticated toolchain would ease the maintenance burden.

### PR DESCRIPTION
**Pull request by exarkun**
Fixes #415 

This adds Dockerfiles for building a flapp server container and a web server container.  It also includes a docker-compose configuration which can launch the two servers and allow them to interact in the necessary ways.  docker-compose is probably not the right tool to use to manage these containers for a production deployment but it is an easy way to be able to see the containers in action.

### CODE REVIEW

File: docker/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    depends_on:
+
+      # Start the flapp server first so the signup.furl is written.
+      # This is subject to startup race-conditions since Docker has no
+      # idea when signup.furl is written: it only knows that it has
+      # _started_ the flapp container.  Experiments suggest the flapp
+      # server typically wins.  Ideally some alternate mechanism could
+      # be used to convey the signup furl between processes (in other
+      # words, fix the application to not use the filesystem for RPC).
+        - "flapp"
```
Comment (meejah) [inline docker/docker-compose.yml]: Yeah :) I've had this "started but not 'ready'" type of problem in lots of situations. Also in the integration tests for tahoe, for example. Unfortunately, I don't have a great solution.

systemd uses a `READY=1` thing, which is remarkably similar to the hack I used in the integration tests (which is to grok stdout for a magic-string). Would love to learn of a better general solution ;)

**Comment by exarkun** [comment]
Thanks!  I think the next step towards getting this deployed is picking a container management tool (Swarm, k8s, etc).  That may suggest some further changes to the Docker work here but since the docker-compose.yml file in this branch demonstrates the code really does start up and work in containers, this seems like a good checkpoint to me.

File: docker/docker-compose.yml (hunk 1 above)
Comment (exarkun) [inline docker/docker-compose.yml]: The systemd `READY=1` example is roughly along the lines of my preferred solution.  Generalizing, the application should emit readiness notification in a way that dependencies can consume it.  Using systemd as a bus for this information seems like it makes sense if systemd is already managing the units.  It presents a uniform interface for delivering the notification and a system for reacting to it (ie, dependent units can start after readiness is signaled).

I'm not sure if there's a way to apply this to a group of Docker-based containers.  But perhaps there's a similar bus available in one of the container management tools.  If not, we could create one ourselves (or we could deliver the notification directly to the web server with no bus, if we don't think we're going to have other use-cases for this feature - and I don't know of any right now).

Stepping back, though, a broader question that I want to discuss is whether the flappserver will continue to be a good solution or if something simpler can get the job done.  I don't know if I'm aware of all of the design considerations that went into choosing the current approach so it'd be great to discuss them before making any major changes.

Comment (meejah) [inline docker/docker-compose.yml]: I don't know why flappserver was used, but I strongly suspect it was along the lines of "we're already using Foolscap, let's use *moar foolscap*" or similar ;) Maybe @daira knows the answer.

**Pull request merged**